### PR TITLE
:sparklings: TK-4757: Log Collector Enhancements

### PR DESCRIPTION
**Pull request by pallav-trilio**

### CODE REVIEW

File: docs/log-collector/README.md [hunk 1]
```diff
@@ -105,6 +108,42 @@ Verify installation with `kubectl tvk-log-collector --help`
 
         kubectl tvk-log-collector --clustered --keep-source-folder --log-level error
 
+- To collect logs by providing gvk :
+
+        kubectl tvk-log-collector --clustered --gvk '[{"group": "<group>", "version": "<version>", "kind": "<kind>"}, {...}]'
```
Comment (msurudwad) [inline docs/log-collector/README.md:113]: Providing JSON string via flags is not user-friendly.
Take input like - 
```
--gvk [Group1/Version1/Kind1, Group2/Version2/Kind2]
```
If found anything easier than this, use that.

File: docs/log-collector/README.md [hunk 2]
```diff
@@ -105,6 +108,42 @@ Verify installation with `kubectl tvk-log-collector --help`
...
+
+- To collect logs by providing label-selector :
+        
+        kubectl tvk-log-collector --clustered  --label-selector '[{"matchLabels":{"key":"value","key2":"value2"}}]'
```
Comment (msurudwad) [inline docs/log-collector/README.md:117]: same here.. Providing JSON string via flags is not user-friendly.
Refer implementation of `kubectl get -l` command.. use the same.

File: tools/log-collector/log_collector.go [hunk 3]
```diff
@@ -30,16 +32,33 @@ import (
 	"github.com/trilioData/tvk-plugins/internal"
 )
 
+type GroupVersionKind struct {
+	Group   string `json:"group"`
+	Version string `json:"version"`
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:37]: add `omitempty`
If `version` not provided, use serverPreferredVesion
If `group` not provided, use "core" grp

File: tools/log-collector/log_collector.go [hunk 4]
```diff
@@ -30,16 +32,33 @@ import (
 	"github.com/trilioData/tvk-plugins/internal"
 )
 
+type GroupVersionKind struct {
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:35]: add validation for this struct - at least `kind` should be given

File: tools/log-collector/log_collector.go [hunk 5]
```diff
@@ -358,31 +377,37 @@ func (l *LogCollector) zipDir() error {
...
+
 	if (!resource.Namespaced && clusteredResources.Has(resource.Kind)) ||
 		(resource.Namespaced && !excludeResources.Has(resource.Kind)) {
 
-		log.Infof("Fetching '%s' Resource", resource.Kind)
-		allObjects = l.getResourceObjects(resourcePath, resource)
-
 		if resource.Name == CRD {
-			allObjects, err = filterTvkSnapshotAndCSICRD(allObjects)
+			log.Infof("Fetching '%s' Resource", resource.Kind)
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:389]: change level to `Debug` .. do the same for other low level information logs.

File: tools/log-collector/log_collector.go [hunk 6]
```diff
@@ -623,3 +659,45 @@ func (l *LogCollector) getResourceEvents(eventResource *apiv1.APIResource, resou
...
+	for i := range labelSelector {
+		ls := labelSelector[i]
+		if MatchLabels(objLabels, ls.MatchLabels) && MatchExpressions(objLabels, ls.MatchExpressions) {
+			return true
+		}
+	}
+	return false
+}
+
+func MatchLabels(podLabels, matchLabels map[string]string) bool {
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:677]: rename `podLabels`

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: cmd/log-collector/cmd/root.go [hunk 7]
```diff
@@ -1,91 +1,99 @@
...
+	namespacesShort            = "n"
+	clusteredUsage             = "specifies clustered object"
+	clusteredDefault           = false
+	clusteredShort             = "c"
+	keepSourceUsage            = "Keep source directory and Zip both"
+	keepSourceDefault          = false
+	keepSourceShort            = "s"
+	configFileUsage            = "specifies the name of the yaml file for inputs to the logCollector flags"
+	configFlagShorthand        = "f"
+	gvkUsage                   = "specifies the json string of all gvk other than log collector handles by default"
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:43]: done

File: cmd/log-collector/cmd/root.go [hunk 8]
```diff
@@ -1,91 +1,99 @@
...
+	gvkFlagShorthand           = "g"
+	labelSelectorUsage         = "specifies the json string of all gvk other than log collector handles by default"
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:45]: done

File: cmd/log-collector/cmd/root.go [hunk 9]
```diff
@@ -105,15 +113,88 @@ func Execute() {
...
 		log.Fatalf("Cannot use flag %s and %s scope at the same time", namespacesFlag, clusteredFlag)
 	}
+
+	if !logCollector.Clustered && len(logCollector.Namespaces) == 0 {
+		log.Fatalf("Either clustered option or at least one namespace in namespace option should be specified for collecting logs")
+	}
 	return nil
 }
+
+func manageFileInputs() (err error) {
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:139]: done

File: docs/log-collector/README.md (hunk 1 above)
Comment (pallav-trilio) [reply docs/log-collector/README.md:113]: done

File: docs/log-collector/README.md (hunk 2 above)
Comment (pallav-trilio) [reply docs/log-collector/README.md:117]: done

File: tools/log-collector/log_collector.go (hunk 3 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:37]: done

File: tools/log-collector/log_collector.go (hunk 4 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:35]: done

File: tools/log-collector/log_collector.go [hunk 10]
```diff
@@ -30,16 +32,33 @@ import (
...
+	CleanOutput      bool     `json:"keep-source-folder"`
+	Clustered        bool     `json:"clustered"`
+	Namespaces       []string `json:"namespaces"`
+	Loglevel         string   `json:"logLevel"`
+	k8sClient        client.Client
+	disClient        *discovery.DiscoveryClient
+	k8sClientSet     *kubernetes.Clientset
+	KubeConfig       string                `json:"kubeConfig"`
+	LabelSelector    []apiv1.LabelSelector `json:"labelSelector,omitempty"`
+	GroupVersionKind []GroupVersionKind    `json:"groupVersionKind"`
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:61]: done

File: tools/log-collector/log_collector.go (hunk 5 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:389]: done

File: tools/log-collector/log_collector.go (hunk 6 above)
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:677]: done

File: tools/log-collector/log_collector.go [hunk 11]
```diff
@@ -623,3 +659,45 @@ func (l *LogCollector) getResourceEvents(eventResource *apiv1.APIResource, resou
...
+			return false
+		}
+		if value != v {
+			return false
+		}
+	}
+	return true
+}
+
+func MatchExpressions(podLabels map[string]string, matchExpr []apiv1.LabelSelectorRequirement) bool {
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply tools/log-collector/log_collector.go:690]: done

File: cmd/log-collector/cmd/root.go [hunk 12]
```diff
@@ -105,15 +113,88 @@ func Execute() {
...
+	}
+}
+
+func parseGVK(gvkstring string) {
+	var gvks []logcollector.GroupVersionKind
+	err := json.Unmarshal([]byte(gvkstring), &gvks)
+	if err != nil {
+		log.Fatalf("Error while parsing GVK : %s", err.Error())
+	}
+	logCollector.GroupVersionKind = gvks
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:190]: done

File: cmd/log-collector/cmd/root.go [hunk 13]
```diff
@@ -105,15 +113,88 @@ func Execute() {
...
+}
+
+func parseLabelSelector(selectorString string) {
+	var lbSelectors []apiv1.LabelSelector
+	err := json.Unmarshal([]byte(selectorString), &lbSelectors)
+	if err != nil {
+		log.Fatalf("Error while parsing label selector : %s", err.Error())
+	}
+	logCollector.LabelSelector = lbSelectors
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:199]: done

File: cmd/log-collector/cmd/root.go [hunk 14]
```diff
@@ -105,15 +113,88 @@ func Execute() {
...
 	}
 	log.SetLevel(level)
 
-	if len(namespaces) != 0 && clustered {
+	if len(logCollector.Namespaces) != 0 && logCollector.Clustered {
 		log.Fatalf("Cannot use flag %s and %s scope at the same time", namespacesFlag, clusteredFlag)
 	}
+
+	if !logCollector.Clustered && len(logCollector.Namespaces) == 0 {
+		log.Fatalf("Either clustered option or at least one namespace in namespace option should be specified for collecting logs")
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply cmd/log-collector/cmd/root.go:134]: done

File: docs/log-collector/README.md [hunk 15]
```diff
@@ -105,6 +108,42 @@ Verify installation with `kubectl tvk-log-collector --help`
...
+      "key2": "value2"
+  - matchLabels:
+      "key3": "value3"
+groupVersionKind:
+  - group: <group>
+    version: <version>
+    kind: <kind>
+  - group: <group2>
+    version: <version2>
+    kind: <kind2>
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply docs/log-collector/README.md]: done

File: docs/log-collector/README.md [hunk 16]
```diff
@@ -105,6 +108,42 @@ Verify installation with `kubectl tvk-log-collector --help`
...
+        
+        kubectl tvk-log-collector --clustered  --label-selector '[{"matchLabels":{"key":"value","key2":"value2"}}]'
+
+- To collect logs by providing config file :
+
+        kubectl tvk-log-collector --config-file <path/to/config/file.yaml>
+
+The format of data in a file should be according to the below example:
+
+```yaml
```
[the remark this answers is not in the document]
Comment (pallav-trilio) [reply docs/log-collector/README.md:125]: done

File: tools/log-collector/log_collector.go [hunk 17]
```diff
@@ -33,9 +34,9 @@ import (
 )
 
 type GroupVersionKind struct {
-	Group   string `json:"group"`
-	Version string `json:"version"`
-	Kind    string `json:"kind"`
+	Group   string `json:"group,omitempty"`
+	Version string `json:"version,omitempty"`
+	Kind    string `json:"kind,omitempty"`
```
Comment (msurudwad) [inline tools/log-collector/log_collector.go:39]: remove `omitempty`

File: cmd/log-collector/cmd/helper.go [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+	var uniqueSelectors []apiv1.LabelSelector
+
+	for idx := range lbSelectors {
+		skip := false
+		for idx1 := range uniqueSelectors {
+			if reflect.DeepEqual(lbSelectors[idx], uniqueSelectors[idx1]) {
+				skip = true
+				break
+			}
+		}
```
Comment (msurudwad) [inline cmd/log-collector/cmd/helper.go:141]: create `set` and use it to check if current gvk/label is already inserted and  `lbSelectors[idx].String()` can give string value to insert into set

File: cmd/log-collector/cmd/helper.go [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+func deDuplicateGVKs(gvks []logcollector.GroupVersionKind) []logcollector.GroupVersionKind {
+	var uniquegvks []logcollector.GroupVersionKind
+	for idx := range gvks {
+		skip := false
+		for idx1 := range uniquegvks {
+			if reflect.DeepEqual(gvks[idx], uniquegvks[idx1]) {
+				skip = true
+				break
+			}
+		}
```
Comment (msurudwad) [inline cmd/log-collector/cmd/helper.go:158]: same here

**Review by msurudwad — APPROVED** [review]

**Pull request merged**